src/gc/proxy.d

```diff
@@ -28,32 +28,67 @@ private
     extern (C) void thread_init();
     extern (C) void thread_term();
 
+    import core.internal.spinlock;
+    static SpinLock instanceLock;
+
     __gshared GC instance;
     __gshared GC proxiedGC; // used to iterate roots of Windows DLLs
-
 }
 
+alias GCInitNoThrow = void function() nothrow @nogc;
 
 extern (C)
 {
 
     void gc_init()
     {
-        config.initialize();
-        ManualGC.initialize(instance);
-        ConservativeGC.initialize(instance);
-        if (instance is null)
-        {
-            import core.stdc.stdio : fprintf, stderr;
-            import core.stdc.stdlib : exit;
+        import core.atomic : atomicLoad, atomicStore;
+
+        auto pinstance = cast(shared(GC)*) &instance;
 
-            fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
-            exit(1);
+        if (atomicLoad(*pinstance) is null)
+        {
+            instanceLock.lock();
+            scope(exit) instanceLock.unlock();
+            {
+                if (atomicLoad(*pinstance) is null)
```

Comment:
I don't think `atomicLoad` is necessary once inside the lock, since the values is guaranteed not to be modified while a thread is inside the lock.

Reply:
I wonder whether you have tried the following trick of using a variable to cache and thus providing as fast, hot-path for the default case:

```
__gshared gcHasBeenInitialized;
if (!gcHasBeenInitialized)
{
 // do your normal locking here
 gcHasBeenInitialized = true;
}
```

TLS variables could be faster, but you could benchmark this for [dynamic modules too](http://david-grs.github.io/tls_performance_overhead_cost_linux/).

Reply:
Would love to see a performance comparison on TLS vs atomic double-checked locking.

Reply:
Double-checked locking should be faster, TLS reads go through an extra indirection and can be very slow if TLS is emulated. Just reading a shared variable by multiple threads doesn't cost much (except maybe for the non-inlined `atomicRead!acq` overhead with dmd).

Reply:
@somzzz let's conservatively leave atomicLoad here and review it in a future pass thx